lib/dataset/groupby.cpp

```diff
@@ -292,28 +332,35 @@ template <class T> struct MakeGroups {
     const auto &values = key.values<T>();
 
     const auto dim = key.dim();
-    std::map<T, GroupByGrouping::group, NanSensitiveLess<T>> indices;
+    std::unordered_map<T, GroupByGrouping::group, std::hash<T>,
+                       nan_sensitive_equal<T>>
+        indices;
     const auto end = values.end();
     scipp::index i = 0;
     for (auto it = values.begin(); it != end;) {
       // Use contiguous (thick) slices if possible to avoid overhead of slice
       // handling in follow-up "apply" steps.
       const auto begin = i;
       const auto &group_value = *it;
-      while (it != end && nan_sensitive_equal(*it, group_value)) {
+      while (it != end && nan_sensitive_equal<T>()(*it, group_value)) {
         ++it;
         ++i;
       }
       indices[group_value].emplace_back(dim, begin, i);
     }
 
-    const Dimensions dims{targetDim, scipp::size(indices)};
     std::vector<T> keys;
     std::vector<GroupByGrouping::group> groups;
-    for (auto &item : indices) {
-      keys.emplace_back(std::move(item.first));
-      groups.emplace_back(std::move(item.second));
-    }
+    keys.reserve(scipp::size(indices));
+    groups.reserve(scipp::size(indices));
+    for (auto &item : indices)
+      keys.emplace_back(item.first);
+    core::parallel::parallel_sort(keys.begin(), keys.end(),
+                                  NanSensitiveLess<T>());
+    for (const auto &k : keys)
+      groups.emplace_back(std::move(indices.at(k)));
```

Comment:
@jl-wynen cppcheck seems to dislike this, can you see why?

```
/home/runner/work/scipp/scipp/lib/dataset/groupby.cpp:361:47: error: Out of bounds access in expression 'indices.at(k)' because 'indices' is empty and 'at' may be non-zero. [containerOutOfBounds]
      groups.emplace_back(std::move(indices.at(k)));
                                              ^
/home/runner/work/scipp/scipp/lib/dataset/groupby.cpp:340:39: note: Assuming condition is false
    for (auto it = values.begin(); it != end;) {
                                      ^
/home/runner/work/scipp/scipp/lib/dataset/groupby.cpp:360:26: note: Assuming container is not empty
    for (const auto &k : keys)
                         ^
/home/runner/work/scipp/scipp/lib/dataset/groupby.cpp:361:47: note: Access out of bounds
      groups.emplace_back(std::move(indices.at(k)));
```